memtester/memtester.8
.TH memtester "8" "November 2024" "memtester 4" "Maintenance Commands"
.SH NAME
memtester \- stress test to find memory subsystem faults.
.SH SYNOPSIS
.B memtester
[\-p \fIPHYSADDR\fR [\-d \fIDEVICE\fR] [\-u]]
<\fIMEMORY\fR>
[\fIITERATIONS\fR]
.SH DESCRIPTION
.\" Add any additional description here
.PP
memtester is an effective userspace tester for stress-testing the memory
subsystem.  It is very effective at finding intermittent and non-deterministic
faults.  Note that problems in other hardware areas (overheating CPU,
out-of-specification power supply, etc.) can cause intermittent memory faults,
so it is still up to you to determine where the fault lies through normal
hardware diagnostic procedures; memtester just helps you determine whether
a problem exists.
.PP
memtester will malloc(3) the amount of memory specified, if possible.  If
this fails, it will decrease the amount of memory requested until it succeeds.
It will then attempt to mlock(3) this memory; if it cannot do so, testing
will be slower and much less effective.  Run memtester as root so that it
can mlock the memory it tests.
.PP
Note that the maximum amount of memory that memtester can test will be less
than the total amount of memory installed in the system; the operating system,
libraries, and other system limits take some of the available memory.
memtester is also limited to the amount of memory available to a single
process; for example, on 32-bit machines with more than 4GB of memory,
memtester is still limited to less than 4GB.
.PP
Note that it is up to you to know how much memory you can safely allocate
for testing.  If you attempt to allocate more memory than is available,
memtester should figure that out, reduce the amount slightly, and try again.
However, this can lead to memtester successfully allocating and mlocking
essentially all free memory on the system -- if other programs are running,
this can lead to excessive swapping and slowing the system down to the point
that it is difficult to use.  If the system allows allocation of more memory
than is actually available (overcommit), it may lead to a deadlock, where
the system halts.  If the system has an out-of-memory process killer (like
Linux), memtester or another process may be killed by the OOM killer.
.PP
So choose wisely.
.PP
.SH OPTIONS
.TP
\-p \fIPHYSADDR\fR
tells memtester to test a specific region of memory starting at physical
address PHYSADDR (given in hex), by mmap(2)ing a device specified by the
-d option (below, or /dev/mem by default).  This is mostly of use to hardware
developers, for testing memory-mapped I/O devices and similar.
Note that the memory region will be overwritten during testing, so it is not
safe to specify memory which is allocated for the system or for other
applications; doing so will cause them to crash.  If you absolutely must test
a particular region of actual physical memory, arrange to have that memory
allocated by your test software, and hold it in this allocated state, then
run memtester on it with this option.
.TP
\-d \fIDEVICE\fR
specifies the device to open with the -p option above.  If not supplied,
/dev/mem is the default.
.TP
\-u\fR
causes memtester to open the device specified by the -d option above without the
O_SYNC flag, which causes the system to use uncached memory semantics on some
systems.
.TP
\fIMEMORY\fR
the amount of memory to allocate and test, in megabytes by default.  You can
include a suffix of B, K, M, or G to indicate bytes, kilobytes, megabytes, or
gigabytes respectively.
.TP
\fIITERATIONS\fR
(optional) number of loops to iterate through.  Default is infinite.
.SH ENVIRONMENT
.PP
If the environment variable MEMTESTER_TEST_MASK is set, memtester treats the
value as a bitmask of which tests (other than the stuck address test) to run.
The value can be specified in decimal, in octal (with a leading 0), or in
hexadecimal (with a leading 0x).  The specific bit values corresponding to
particular tests may change from release to release; consult the list of tests
in the source for the appropriate index values for the version of memtester you
are running.  Note that skipping some tests will reduce the time it takes for
memtester to run, but also reduce memtester's effectiveness.
.SH NOTE
.PP
memtester must be run with root privileges to mlock(3) its pages.  Testing
memory without locking the pages in place is mostly pointless and slow.
.SH EXIT CODE
.PP
memtester's exit code is 0 when everything works properly.  Otherwise,
it is the logical OR of the following values:
.TP
\f0x01
error allocating or locking memory, or invocation error
.TP
\f0x02
error during stuck address test
.TP
\f0x04
error during one of the other tests
.SH AUTHOR
Written by Charles Cazabon.
.SH "REPORTING BUGS"
Report bugs to <charlesc-memtester-bugs@pyropus.ca>.
.PP
.SH COPYRIGHT
Copyright \(co 2001-2024 Charles Cazabon
.br
This is free software; see the file COPYING for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


memtester/CHANGELOG
Version 4.7.1
16 December 2024
  -fix compile error with stricter C23/gcc-15 due to incompatible type declaration.
  Thanks: Rudi Heitbaum.

Version 4.7.0
28 November 2024
  -fix alignment issue causing crash when specifying a memory amount to test
  that is more than the minimum 1 page of memory, but less than 2 full pages.
  Thanks: Erwan Szymanski.  Also reported by Chenghan Lee and K R Pallavi.
  -add `-u` option to allow uncached memory semantics on some systems when
  using the -p option.  Thanks: Florian Fainelli.
  -fix manpage highlighting of args.  Thanks: László Böszörményi.
  -add missing option documentation to man page.

Version 4.6.0
19 December 2022
  -fix warning on signedness of the ONE constant.  Thanks: Marek Vasut.
  -add note to conf-cc about how to silence gcc's incorrect warning
  "suggest parentheses around assignment".  Thanks: Marek Vasut.
  -fix overflow in output for reported address.  Thanks: Marek Vasut.
  -add note about randomizing values used at runtime by seeding PRNG.
  Disabled by default for reproducibility.

Version 4.5.1
30 May 2021
  -documentation-only update.
  -add note to README about build problem with Tru64, with workaround.
  Thanks: Víctor Ostorga.

Version 4.5.0
20 September 2020
  -add note to README about problems with HP/UX prior to v.11.  Change code
  so that buffer still gets aligned even if you comment out the use of mlock()
  on such a platform.  Thanks: Keith Monahan.
  -fix clear-buffer using incorrect param, causing spurious test failures.
  Thanks: Debin Dong.
  -fix shell-ism in Makefile.  Thanks: Fabrice Fontaine.

Version 4.3.0
9 June 2012
  -add ability to mmap a device other than /dev/mem, enabling easier testing
  of memory for hardware engineers et al if their system's kernel exports the
  memory they're interested in via /dev/mydevice or similar.  Thanks:
  Jean-Noël Avila.
  -add ability to run only specified tests by setting the environment
  variable MEMTESTER_TEST_MASK to a bitmask of the indexes of the tests to be
  run.  Thanks: Ian Alexander.

Version 4.2.2
22 July 2011
  -add progress message for one more errno value (EAGAIN) in response to failed
  mlock; BSDs use this to indicate the lock failed due to being over a system
  or process limit, much like ENOMEM.

Version 4.2.1
3 October 2010
  -fix offsets/addresses were not being reported correctly in test failure
  messages.  Thanks: Anton Sergeev.

Version 4.2.0
30 July 2010
  -define _FILE_OFFSET_BITS to 64 by default in conf-cc, which causes some
  32-bit systems with larger-than-32-bit address spaces to have a 64-bit off_t,
  allowing testing of larger chunks of memory.  Thanks to Steven A. Falco for
  the suggestion.  Let me know if this definition causes problems for anyone.
  -add tests of 8-bit-wide and 16-bit-wide random writes, to enable verifying
  the correct operation of hardware.  Thanks: Dick Hollenbeck.  If these tests
  trigger unaligned access problems on your platform, you can eliminate these
  tests by removing the -DTEST_NARROW_WRITES definition from the conf-cc file.

Version 4.1.3
28 February 2010
  -fix 64-bit data patterns with some versions of gcc.  Thanks: Tony Battersby.
  -clarify `make install` in readme.  Thanks: Marc Lobelle.

Version 4.1.2
28 July 2009
  -fix portability issue with MAP_LOCKED flag.  Thanks: Scott Haneda.
  -remove debugging output accidentally left in v.4.1.0.
  -cleanups

Version 4.1.1
24 July 2009

  -memtester.h was missing from the 4.1.0 tarball; release update to fix that.
  Thanks: Owen Leonard.

Version 4.1.0
23 July 2009

  -added the ability to test a specific physical region of memory (by mmap'ing
  /dev/mem) with the new -p option, which takes a hex starting address as a
  value.  This is mostly of use to developers trying to verify memory or I/O
  mapped devices (on an embedded system, for instance). Thanks: Allon Stern.
  -re-add the ability to set a suffix on the memory to allocate: "3G", "128K", etc,
  mostly for use with the above new feature, where the "memory" to be tested is
  less than a megabyte, but also useful for users wanting to test many gigabytes;
  you no longer have to do the conversion-to-megs in your head.
  -documentation updates and clarifications.

Version 4.0.8
21 November 2007

  -add a startup check for amount of memory being greater than the possible
  address space; prevents user confusion on 32-bit systems that use addressing
  tricks to have >4GB total system memory.  Thanks: Michael Kelly.
  -documentation updates

Version 4.0.7
13 May 2007

  -fix a bug in the align-to-page logic which may have prevented memtester
  from mlock()ing the memory it was trying to test on some systems.
  Thanks: Baif Chen.

Version 4.0.6
15 November 2006

  -test algorithm improvement: the walking 0 bits test was only walking
  the 0 bit in one direction, instead of walking it up and back down
  the line the way it was intended to.  Thanks: Tim Rule.
  -formatting cleanups.

Version 4.0.5
10 March 2005

  -change to the method of allocating and locking memory; if we get EPERM
  when trying to mlock(), reset the amount of memory desired to the original
  amount and try again without mlock().  The reason for this is that on many
  systems, mlock() won't return EPERM until after having returned ENOMEM for
  a larger amount.  The new behaviour allows processes without mlock privileges
  to still test the fully-specified amount or as much memory as can be
  allocated.  Thanks for the suggestion and testing to Dan Bradley.

Version 4.0.4
26 August 2004

  -make memtester's exit code meaningful.  See the manpage for its meaning.
  Thanks to Wurzel Parsons-Keir, who sent a patch for the code, so I only had
  to document it.

Version 4.0.3
10 August 2004

  -small changes to enable building with dietlibc and a few other environments
  that don't even attempt to provide the various Posix definitions.
  -cosmetic fixes to output.
  -restore the reduce-and-retry loop of memory locking from version 2.


memtester/mcu/memtester.c
/*
 * memtester version 4
 *
 * Very simple but very effective user-space memory tester.
 * Originally by Simon Kirby <sim@stormix.com> <sim@neato.org>
 * Version 2 by Charles Cazabon <charlesc-memtester@pyropus.ca>
 * Version 3 not publicly released.
 * Version 4 rewrite:
 * Copyright (C) 2004-2024 Charles Cazabon <charlesc-memtester@pyropus.ca>
 * Licensed under the terms of the GNU General Public License version 2 (only).
 * See the file COPYING for details.
 *
 */

#define __version__ "4.7.1"

#include <stddef.h>
#include <stdlib.h>
#include <stdio.h>
#include <time.h>
#include <errno.h>

#include "memtester.h"
#include "tests.h"

#define EXIT_FAIL_NONSTARTER    0x01
#define EXIT_FAIL_ADDRESSLINES  0x02
#define EXIT_FAIL_OTHERTEST     0x04

struct test tests[] = {
    { "Random Value", test_random_value },
    { "Compare XOR", test_xor_comparison },
    { "Compare SUB", test_sub_comparison },
    { "Compare MUL", test_mul_comparison },
    { "Compare DIV",test_div_comparison },
    { "Compare OR", test_or_comparison },
    { "Compare AND", test_and_comparison },
    { "Sequential Increment", test_seqinc_comparison },
    { "Solid Bits", test_solidbits_comparison },
    { "Block Sequential", test_blockseq_comparison },
    { "Checkerboard", test_checkerboard_comparison },
    { "Bit Spread", test_bitspread_comparison },
    { "Bit Flip", test_bitflip_comparison },
    { "Walking Ones", test_walkbits1_comparison },
    { "Walking Zeroes", test_walkbits0_comparison },
#ifdef TEST_NARROW_WRITES
    { "8-bit Writes", test_8bit_wide_random },
    { "16-bit Writes", test_16bit_wide_random },
#endif
    { NULL, NULL }
};


/*
 * arg1: Test start ADDR
 * arg2: Test size (Bytes)
 * arg3: Test total loops
 */
int memtester(ul test_start_physaddr, size_t test_size, ul loops)
{
    ul loop, i;
    size_t bufsize, halflen, count;
    ulv *buf;
    ulv *bufa, *bufb;
    int exit_code = 0;

    printf("memtester version " __version__ " (%d-bit)\n", UL_LEN);
    printf("Copyright (C) 2001-2024 Charles Cazabon.\n");
    printf("Licensed under the GNU General Public License version 2 (only).\n");
    printf("\n");
    printf("Warning: Ensure you have closed Dcache!!\n");
    printf("\nMemtest start ADDR: " FMT_TARGET ", Test Size is 0x%lx Bytes, Test loops is %lu\n", test_start_physaddr,  test_size, loops);

    /* In theory, you might be able to get slightly better error detection if you randomly seed
       the pseudo-random number generator and run memtester multiple times in sequence.
       However, that benefit is probably very, very slight and won't matter.  Leaving it unseeded
       results in a constant seed, so exactly the same values are used test-to-test, giving
       better reproducibility, so this is disabled by default.  You can uncomment it to enable.
       Note there are no security implications here */
    srand(1);

    buf = (ulv *)test_start_physaddr;
    bufsize = test_size;

    halflen = bufsize / 2;
    count = halflen / sizeof(ul);
    bufa = (ulv *) buf;
    bufb = (ulv *) ((size_t) buf + halflen);

    for(loop=1; ((!loops) || loop <= loops); loop++) {
        printf("Loop %lu", loop);
        if (loops) {
            printf("/%lu", loops);
        }
        printf(":\n");
        printf("  %-20s: ", "Stuck Address");
        fflush(stdout);
        if (!test_stuck_address(buf, bufsize / sizeof(ul))) {
             printf("ok\n");
        } else {
            exit_code |= EXIT_FAIL_ADDRESSLINES;
        }
        for (i=0;;i++) {
            if (!tests[i].name) break;
            printf("  %-20s: ", tests[i].name);
            if (!tests[i].fp(bufa, bufb, count)) {
                printf("ok\n");
            } else {
                exit_code |= EXIT_FAIL_OTHERTEST;
            }
            fflush(stdout);
        }
        printf("\n");
        fflush(stdout);
        if (exit_code) break;
    }

    if (exit_code) {
        printf("Mem Test Fail!, exit_code = %d\n", exit_code);
        printf("Fail, please check!!\n");
        fflush(stdout);
    } else {
        printf("Mem Test Pass!\n");
        printf("OK.\n");
        fflush(stdout);
    }
    
    return 0;
}
